### PR TITLE
Patched Wigner3j not working with float that were half integers

### LINKED ISSUE #26219: Wigner3j has problems with Half Integers
I was trying to find the result of this operation and am getting this error for half integers
```
>>> w3j = Wigner3j(0,0,1/2,1/2,1/2,-1/2)    
>>> w3j.doit()
Traceback (most recent call last):
  File "<console>", line 1, in <module>
  File "D:\Orgs\sympy\sympy\physics\quantum\cg.py", line 158, in doit 
    return wigner_3j(self.j1, self.j2, self.j3, self.m1, self.m2, self.m3)
  File "D:\Orgs\sympy\sympy\physics\wigner.py", line 199, in wigner_3j
    raise ValueError("j values must be integer or half integer")      
ValueError: j values must be integer or half integer
>>> w3j = Wigner3j(0.5,0.5,0.5,0.5,0.5,0.5)
>>> w3j.doit()
Traceback (most recent call last):
  File "<console>", line 1, in <module>
  File "D:\Orgs\sympy\sympy\physics\quantum\cg.py", line 158, in doit 
    return wigner_3j(self.j1, self.j2, self.j3, self.m1, self.m2, self.m3)
  File "D:\Orgs\sympy\sympy\physics\wigner.py", line 199, in wigner_3j    raise ValueError("j values must be integer or half integer")      
ValueError: j values must be integer or half integer
```
It is working fine for integer values but gives an error of values with half integers .

### PR DESCRIPTION
**Pull request by T-vaccari**
This pull request addresses an issue where the Wigner3j function would raise a ValueError when encountering half-integer inputs recognized as floats. The problem stemmed from incomplete validation checks, which failed to properly handle such inputs.

#### References to other Issues or PRs
Fixed #26219

#### Brief description of what is fixed or changed
To resolve this issue, I implemented enhanced validation checks within the function. Now, all inputs are meticulously validated to ensure they are either integers or half-integers. This modification ensures robustness and prevents the function from failing due to invalid input types.

NO ENTRY

### CODE REVIEW

File: sympy/physics/wigner.py [hunk 1]
```diff
@@ -107,7 +107,7 @@ def _calc_factlist(nn):
             _Factlist.append(_Factlist[ii - 1] * ii)
     return _Factlist[:int(nn) + 1]
 
-
+from fractions import Fraction
```
[the remark this answers is not in the document]
Comment (T-vaccari) [reply sympy/physics/wigner.py:110]: It's an error I used it to check something

**Comment by T-vaccari** [comment]
Why it's not working without importing fractions

File: sympy/physics/wigner.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (VascoSch92) [reply sympy/physics/wigner.py:110]: Ok but why are you importing here and not at the top of the script? Is there a reason?

[the remark this answers is not in the document]
Comment (T-vaccari) [reply sympy/physics/wigner.py:110]: No there's no reason , I can modify it if you want. I want to say that this is my first pull request so I am open to receive every type of advice

File: sympy/physics/wigner.py [hunk 2]
```diff
@@ -194,12 +194,34 @@ def wigner_3j(j_1, j_2, j_3, m_1, m_2, m_3):
...
-            int(j_3 * 2) != j_3 * 2:
+
+    def convert_float_to_rational_if_half_integer(value):
+        if isinstance(value, float):
+            fraction_value = Fraction(value).limit_denominator()
+            if fraction_value.denominator == 2:
+                return fraction_value
+        return value
+
+    # Example list of values
```
Comment (VascoSch92) [inline sympy/physics/wigner.py:205]: why `Example`? These are the inputs of the method and not an example :-)

File: sympy/physics/wigner.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (T-vaccari) [reply sympy/physics/wigner.py:110]: Thank you for your feedback, @VascoSch92. I appreciate your review. I'll implement your suggestion by moving the import statement to the beginning of the script for better clarity and conformity with best practices. Additionally, I'll rerun the test to verify if the changes resolve the issue. If there are any further concerns or recommendations, please don't hesitate to share them.

File: sympy/physics/wigner.py (hunk 2 above)
Comment (T-vaccari) [reply sympy/physics/wigner.py:205]: Yes I left the comment while I was testing the code for myself using example values , my bad sorry

File: sympy/physics/wigner.py [hunk 3]
```diff
@@ -194,12 +194,34 @@ def wigner_3j(j_1, j_2, j_3, m_1, m_2, m_3):
 
     - Jens Rasch (2009-03-24): initial version
     """
-    if int(j_1 * 2) != j_1 * 2 or int(j_2 * 2) != j_2 * 2 or \
-            int(j_3 * 2) != j_3 * 2:
+
+    def convert_float_to_rational_if_half_integer(value):
```
[the remark this answers is not in the document]
Comment (T-vaccari) [reply sympy/physics/wigner.py:198]: No, I think it's not going to be used elsewhere

**Comment by T-vaccari** [comment]
Someone knows why it's not passing benchmark test  ?

**Comment by shishir-11** [comment]
> Someone knows why it's not passing benchmark test ?

You should check out some of the latest PRs in the repository . It has been a regular occurrence , I don't know why it is happening , but it's not about your PR.

**Comment by T-vaccari** [comment]
> > Someone knows why it's not passing benchmark test ?
> 
> You should check out some of the latest PRs in the repository . It has been a regular occurrence , I don't know why it is happening , but it's not about your PR.

Thank you , I'm going to check other PR to see if I can discover why it's not passing it.

**Comment by T-vaccari** [comment]
Is this pull request ok or I have to change something?

File: sympy/physics/wigner.py [hunk 4]
```diff
@@ -69,7 +69,7 @@
 from sympy.matrices.dense import zeros
 from sympy.matrices.immutable import ImmutableMatrix
 from sympy.utilities.misc import as_int
-
+from fractions import Fraction
```
Comment (oscarbenjamin) [inline sympy/physics/wigner.py:72]: We don't want to use Fraction. SymPy has its own versions of these things.

Comment (T-vaccari) [reply sympy/physics/wigner.py]: My bad, I didn't think about using its own module for fractions,  now I have corrected it .  Let me know what do you think about this changes !

**Comment by abhiphile** [comment]
Can you write some tests for your PR.

**Comment by T-vaccari** [comment]
> Can you write some tests for your PR.

I've already added two test for this pull requests , if you have in mind something else let me know!

File: sympy/physics/wigner.py [hunk 5]
```diff
@@ -194,12 +194,28 @@ def wigner_3j(j_1, j_2, j_3, m_1, m_2, m_3):
...
     """
-    if int(j_1 * 2) != j_1 * 2 or int(j_2 * 2) != j_2 * 2 or \
-            int(j_3 * 2) != j_3 * 2:
+
+    def convert_float_to_rational_if_half_integer(value):
+        if isinstance(value, float):
+            fraction_value = Rational(value).limit_denominator()
+            if fraction_value.denominator == 2:
+                return fraction_value
+        return value
```
Comment (smichr) [inline sympy/physics/wigner.py:203]:
```suggestion
        if isinstance(value, float):
            if equal_valued((v:=2*value), (i:=int(v))):
                return Rational(i, 2)
        return value
```
`equal_valued` can be imported from sympy.core.numbers. Should this be passing if someone passes 3/3?

Comment (T-vaccari) [reply sympy/physics/wigner.py:203]: Yes it should pass because it's an integer

Comment (oscarbenjamin) [reply sympy/physics/wigner.py:203]: We don't want to use `limit_denominator` for this.

**Comment by T-vaccari** [comment]
I think the code quality check is bugged because its giving error of whitespace in line written without problems

File: sympy/physics/wigner.py (hunk 5 above)
Comment (T-vaccari) [reply sympy/physics/wigner.py]: In my machine it's not failing the code quality but on git it's failing it for `equal_valued` why ?

File: sympy/physics/wigner.py [hunk 6]
```diff
@@ -69,7 +69,7 @@
 from sympy.matrices.dense import zeros
 from sympy.matrices.immutable import ImmutableMatrix
 from sympy.utilities.misc import as_int
-
+from sympy import Rational
```
Comment (smichr) [inline sympy/physics/wigner.py:72]:
```suggestion
from sympy.core.numbers import Rational, Integer, equal_valued
```

File: sympy/physics/wigner.py [hunk 7]
```diff
@@ -194,12 +194,28 @@ def wigner_3j(j_1, j_2, j_3, m_1, m_2, m_3):
 
     - Jens Rasch (2009-03-24): initial version
     """
-    if int(j_1 * 2) != j_1 * 2 or int(j_2 * 2) != j_2 * 2 or \
-            int(j_3 * 2) != j_3 * 2:
+    def convert_float_to_rational_if_half_integer(value):
+        if isinstance(value, float):
+            if (value.is_integer()):
+                return Rational(int(value))
```
Comment (smichr) [inline sympy/physics/wigner.py:200]:
```suggestion
                return Integer(int(value))
```

**Comment by T-vaccari** [comment]
Thank you very much, @smichr . I would greatly appreciate it if you could explain why you imported those modules from the core. Additionally, I'm curious about how you made a commit on my pull request. I'm not very experienced with Git, and I didn't realize that I could make a commit to an external pull request. Thank you for your help!

**Comment by smichr** [comment]
It seems like best practice to import directly from the module (otherwise the entire namespace of the module from which you import is being added to the local namespace?).  

The commit was made as part of the review process. Since I am reviewing your work and know the change that needs to take place, I will sometimes just make the change to save another commit from you and review from me. I think I am able to make the commit (instead of only propose the commit) because I am a maintainer.

**Comment by smichr** [comment]
I am not sure if it is a good idea to convert the input to Integer/Rational. Do you see an issue with doing so? In general, calculations with Python base types is better than SymPy Numbers as long as there are no precision issues.

File: sympy/physics/wigner.py [hunk 8]
```diff
@@ -56,7 +56,7 @@
 from sympy.core.add import Add
 from sympy.core.numbers import int_valued
 from sympy.core.function import Function
-from sympy.core.numbers import (I, Integer, pi)
+from sympy.core.numbers import (I, Integer, pi, Rational, equal_valued)
```
Comment (smichr) [inline sympy/physics/wigner.py:59]:
```suggestion
from sympy.core.numbers import (Float, I, Integer, pi, Rational, equal_valued)
```

File: sympy/physics/wigner.py [hunk 9]
```diff
@@ -108,6 +108,19 @@ def _calc_factlist(nn):
...
+def _Integer_or_halfInteger(value):
+    if isinstance(value, int):
+        return Integer(value)
+    elif isinstance(value, (float, Float)):
+        if isinstance(value, float) and value.is_integer():
+            return Integer(int(value))
+        elif (equal_valued((v:=2*value), (i:=int(v)))):
+            return Rational(i, 2)
+    elif isinstance(value, Integer):
+        return value
```
Comment (smichr) [inline sympy/physics/wigner.py:120]:
```suggestion
    elif isinstance(value, (float, Float)):
        if isinstance(value, float) and value.is_integer():
            return Integer(int(value))
        elif (equal_valued((v:=2*value), (i:=int(v)))):
            return Rational(i, 2)
    elif isinstance(value, Integer):
        return value
    elif isinstance(value, Rational) and value.q == 2:
        return value
```

**Comment by T-vaccari** [comment]
> I am not sure if it is a good idea to convert the input to Integer/Rational. Do you see an issue with doing so? In general, calculations with Python base types is better than SymPy Numbers as long as there are no precision issues.

I think it's not a problem to convert the input to integer/Rational. However, we should keep an eye on performance since operations with Python types are generally slightly faster than those with SymPy types. As long as there are no precision issues, using Python base types for calculations should be fine.

**Comment by oscarbenjamin** [comment]
This seems to have caused a regression: gh-27012

**Comment by T-vaccari** [comment]
> This seems to have caused a regression: [gh-27012](https://github.com/sympy/sympy/issues/27012)

Hy @oscarbenjamin , now I try to understand what’s the problem

**Pull request merged**